### ISSUE TITLE
[RFE] Exclude Directories in Autoloading

### ISSUE THREAD

**Issue by rfludwick** [composer/composer#4456]
I could really, really use the ability to exclude certain directories/files in the autoloading definition in `composer.json`.

The gist is this: my workplace uses satis for some internal repos. Some devs who are no longer with the organization ended up duplicating several classnames. This isn't an issue for a couple of our departments who actually *don't* use autoloading, but for those of us who do, myself included, it's a major pain in the ass.

Basically, I get alerts on every composer update/install about multiple classes found, where composer loads the first one it found. I would simply just include the directories I want out of the repo, but there are 40 directories in there. And I only want to exclude one of them.

tl;dr; give me the ability to exclude a directory or filename, ex:

```json
"autoload": {
  "classmap": [
    "vendor/something",
    "!vendor/something/bad"
  ]
}
```

**Comment by jakoch** [comment]
The real solution to this problem is to fix the duplicated class names either by renaming them or better by introducing or using more fine-grained namespaces. You could work through your packages and add namespaces per package (`\vendor\package-1\a` & `\vendor\package-2\a`) and then simply switch to PSR-4.

But this seems to be one of the deployment situations where one might be running composer in two stages. Please let me suggest to simply add a cleanup step between pulling dependencies and generating the Composer Autoloader files, like so:

1. Do not generate files for Composer Autoloader on install or update by using `--no-autoloader`
    - `php composer.phar install --no-autolader`
2. Cleanup the mess: delete the folders with the duplicated classes
3. Generate the autoloader files
   - `php composer.phar dump-autoload --optimize`

(You could also try to use `$loader = require __DIR__ . '/vendor/autoload.php';` and hack your way up to the `$classMap` to modify it. Or manipulate the generated classmap file directly after generation and before its used by the Autoloader. Or you could skip using/including the Composer Autoloader and roll your own, probably depending on the Composer generated files, but taking excludes into account. Lots of options... :)

**Comment by rfludwick** [comment]
I don't have control over the libraries, nor can I force them to fix the problem. And I need them. However, the three-step course you've suggested will probably work just fine for me. Thanks for that suggestion!

**Comment by rquadling** [comment]
If you script your cleanup, you could run that as part of a post-install / post-update script.

And then also include the dump-autoload --optimize.

e.g.
```
    "scripts": {
        "post-install-cmd": [
            "php ./scripts/composerCleanup.php",
            "composer dump-autoload --optimize"
        ],
        "post-update-cmd": [
            "php ./scripts/composerCleanup.php",
            "composer dump-autoload --optimize"
        ]
    }
```

You would have to remember to use ```--no-autoloader``` every time (unless there's a way to configure that somewhere?).

**Comment by stof** [comment]
The other alternative is to accept to see this message in your output until the day the libraries are fixed, assuming that the first class definition is the one you need. This is just a warning message in the output, but it does not make Composer fail.

**Comment by Seldaek** [comment]
This is now possible with https://getcomposer.org/doc/04-schema.md#exclude-files-from-classmaps
